### PR TITLE
Graph colouring cluster ids

### PR DESCRIPTION
**Pull request by mundya**
Cluster IDs are used to ensure the multicast nets representing signals which originate from multiple chips can be differentiated if they take different routes at the same router. Previously each chip was assigned a unique identifier but this is unnecessary and wastes both routing table entries and keyspace.

This PR uses graph colouring to reduce the number of different cluster IDs applied, with the result the fewer bits in the key and fewer routing tables are used to simulate the same model.

### CODE REVIEW

File: nengo_spinnaker/netlist/key_allocation.py [hunk 1]
```diff
@@ -106,26 +106,163 @@ def build_mn_net_graph(nets_routes, prior_constraints=None):
...
+        signal_routes = operators_signal_routes[operator]
+        n_clusters = len({placements[vx] for vx in vertices})
+
+        if len(signal_routes) == 0 or n_clusters == 1:
+            # If the operator has no outgoing signals, or only one cluster,
+            # then assign the same identifier to all of the vertices.
+            for vertex in vertices:
+                vertex.cluster = 0
+        else:
+            # Otherwise try to allocate a few cluster IDs as are required to
```
Comment (mundya) [inline nengo_spinnaker/netlist/key_allocation.py]: "Otherwise try to allocate _as_ few"

File: nengo_spinnaker/netlist/key_allocation.py [hunk 2]
```diff
@@ -106,26 +106,163 @@ def build_mn_net_graph(nets_routes, prior_constraints=None):
...
+            cluster_graph[source]  # Ensure every cluster is in the graph
+
+            # Traverse the tree to build the routing directions taken by the
+            # signals.
+            for _, chip, routes in tree.traverse():
+                rs = frozenset(routes)
+                chips_signals_routes_clusters[chip][signal][rs].add(source)
+
+    # Build the graph representing the constraints on cluster identities by
+    # iterating over this new data.
```
Comment (mundya) [inline nengo_spinnaker/netlist/key_allocation.py]: "~~by iterating...~~"

File: nengo_spinnaker/builder/builder.py [hunk 3]
```diff
@@ -340,8 +339,10 @@ def make_netlist(self, *args, **kwargs):
                 self, *args, **kwargs
             )
 
-            operator_vertices[op] = vxs
-            vertices.append(vxs)
+            if isinstance(vxs, (list, tuple, set)):
```
Comment (neworderofjamie) [inline nengo_spinnaker/builder/builder.py]: This is unpleasant - could operators whose `make_vertices` method returns a single vertex do so wrapped in a tuple?

**Review by neworderofjamie — COMMENTED** [review]
Looks good to me - nice re-use of colour_graph

**Pull request merged**